client/db/bolt/db.go

```diff
@@ -163,14 +164,55 @@ func NewDB(dbPath string, logger dex.Logger) (dexdb.DB, error) {
 	return bdb, bdb.upgradeDB()
 }
 
+func (db *BoltDB) fileSize(path string) int64 {
+	stat, err := os.Stat(path)
+	if err != nil {
+		db.log.Errorf("Failed to stat %v: %v", path, err)
+		return 0
+	}
+	return stat.Size()
+}
+
 // Run waits for context cancellation and closes the database.
 func (db *BoltDB) Run(ctx context.Context) {
-	<-ctx.Done()
+	<-ctx.Done() // wait for shutdown to backup and compact
+
+	// Create a backup in the backups folder.
+	db.log.Infof("Backing up database...")
 	err := db.Backup()
 	if err != nil {
-		db.log.Errorf("unable to backup database: %v", err)
+		db.log.Errorf("Unable to backup database: %v", err)
+	}
+
+	// Only compact the current DB file if there are excessive free pages.
+	if db.Stats().FreePageN < 32 { // 128 KiB for 4096 page size
+		db.Close()
+		return
+	}
+
+	// Compact the database by writing into a temporary file, closing the source
+	// DB, and overwriting the original with the compacted temporary file.
+	db.log.Infof("Compacting database...")
+	srcPath := db.Path()                    // before db.Close
+	compFile := srcPath + ".tmp"            // deterministic on *same fs*
+	err = db.BackupTo(compFile, true, true) // overwrite and compact
```

Comment:
Would it be faster to compact the `Backup` and just copy it back?

Reply:
I guess if the compaction went bad, you wouldn't have a database left.

Reply:
We had the same thought it seems.  It's what I was getting at in https://github.com/decred/dcrdex/pull/1486#discussion_r814929231.  If we never see trouble with the compacted DBs when we restart dexc, I suppose we can declare it safe and do it directly.

Reply:
Okay as is.